File: Clean-Architecture/Feature/dart/example/test/features/notifications/domain/usecases/send_notification_test.dart
import 'package:clean_architecture_feature_based_flutter_example/core/result/result.dart';
import 'package:clean_architecture_feature_based_flutter_example/features/notifications/domain/entities/notification_request.dart';
import 'package:clean_architecture_feature_based_flutter_example/features/notifications/domain/repositories/notification_repository.dart';
import 'package:clean_architecture_feature_based_flutter_example/features/notifications/domain/usecases/send_notification.dart';
import 'package:flutter_test/flutter_test.dart';

class FakeNotificationRepository implements NotificationRepository {
  NotificationRequest? captured;

  @override
  Future<Result<SentNotification>> send(NotificationRequest request) async {
    captured = request;
    return const Success(SentNotification(id: '1', channel: 'email', status: 'sent'));
  }
}

void main() {
  test('SendNotification sends valid requests through the contract', () async {
    final repository = FakeNotificationRepository();
    final useCase = SendNotification(repository);
    final result = await useCase(const NotificationRequest(
      channel: 'email',
      recipient: 'student@example.com',
      message: 'Welcome',
    ));

    expect(repository.captured?.recipient, 'student@example.com');
    expect(result, isA<Success<SentNotification>>());
  });
}

File: Clean-Architecture/Feature/Javascript/example/test/notifications/send-notification.test.js
import test from 'node:test';
import assert from 'node:assert/strict';
import { SendNotification } from '../../src/features/notifications/domain/usecases/SendNotification.js';

class FakeNotificationGateway {
  requests = [];
  async send(request) {
    this.requests.push(request);
    return { ...request, id: 'n-1', status: 'sent' };
  }
}

test('SendNotification delegates normalized data to the gateway', async () => {
  const gateway = new FakeNotificationGateway();
  const useCase = new SendNotification({ notificationGateway: gateway });
  const result = await useCase.execute({
    channel: ' EMAIL ',
    recipient: ' student@example.com ',
    message: ' Welcome ',
  });
  assert.equal(result.channel, 'email');
  assert.equal(result.recipient, 'student@example.com');
  assert.equal(result.message, 'Welcome');
  assert.equal(gateway.requests.length, 1);
});
